### DISCUSSION TITLE
🚀 A better way to get values of custom fields via EE model system

### DISCUSSION THREAD

**Discussion by jcogs-design** [ExpressionEngine/ExpressionEngine#4015] [Feature Requests]
### Suggest an idea for ExpressionEngine

**Is your feature request related to a problem? Please describe.**
I noticed today that most of my add-ons have somewhere within them a couple of functions defined whose sole purpose is to facilitate the retrieval of custom field values for entries. The steps are:

- use a model request to get the "id" of the field table containing the custom field data you want to look up
- use a different model request to recover the value of the custom field 

I end up with stuff like this ... 
```
    /**
     * Get ID for a named EE Field
     *
     * @param  string $key
     * @return int|boolean 
     */
    public function get_ee_field($key)
    {
        if ($key && $field_id = ee('Model')->get('ChannelField')->filter('field_name', $key)->first()->field_id) {
            return $field_id;
        } else {
            return false;
        }
        // Get the field ID of the $key
    }
```
and this ... 
```
ee('Model')->get('ChannelEntry')->filter('entry_id', $entry_id)->first()->{'field_id_' . $field_id}
```

And don't get me started about what you have to do if you want / need to know the 'default_field_value' of a custom field ... a non trivial exercise.

Accessing custom field values is something that would / should be better handled within EE itself.

The documentation for doing this [in the docs](https://docs.expressionengine.com/latest/development/models/channel-entry.html#get-an-entrys-custom-field) is concise, but disingenously leaves out discussing how you determine what the field_id for a custom field is, or how you find out what the default value is. 

**Describe the solution you'd like**

- For the EE model / ChannelEntry system to include the ability to retrieve custom field values based on the name of the custom field or the ID of the custom field.
- For EE to optionally return the field default value for a custom field where the value is not set in the relevant EE table

Something like ... 
```
$value = ee('Model')->get('ChannelEntry')
                ->filter('entry_id', $entry_id)
                ->custom_field($field_name or $field_id, $return_default_value_if_not_set_bool);
```

**Describe alternatives you've considered**
Doing it manually as described in intro... 

**Teachability, Documentation, Adoption, Migration Strategy**
This would be unlikely to replace anything existing in EE, so all that would be needed would be documentation on the new method / function for the ChannelEntry model. The method would of course not be backward compatible, so probably needs to be linked to a minor EE version bump (e.g. 7.4?) so add-on developers can put in fall-back logic if their add-ons will possibly cross the boundary between before / after etc.

- [ ] I am capable and would like to work on implementation of this feature if it is considered.

**Comment by intoeetive** [comment]
@jcogs-design you are probably looking for `$entry->getCustomField($name)` method, which returns field facade... and which has zero documentation :)
